04_condition_variable/exercises/03a_ping_pong.cpp

```diff
@@ -1,26 +1,92 @@
+#include <chrono>
+#include <condition_variable>
+#include <cstdlib>
 #include <iostream>
+#include <mutex>
+#include <sstream>
 #include <thread>
-#include <chrono>
+
 using namespace std;
 
 class PingPong {
     int repetitions_;
+    int repetitionsCounter_;
+    mutex coutMutex_;
+    condition_variable pingOrPong_;
+    bool pingTurn_;
 
 public:
     PingPong(int repetitions)
-        : repetitions_(repetitions)
+        : repetitions_(repetitions), repetitionsCounter_(1), pingTurn_(false)
     {}
 
+    void safePongExit(stringstream& ss) {
+        lock_guard<mutex> lockCout(coutMutex_);
+        ss << "Repetitions end!\n";
+        cout << ss.rdbuf();
+        exit(0);
+    }
+    
+    void safeStopExit() {
+        lock_guard<mutex> lockCout(coutMutex_);
+        stringstream ss;
+        
+        if(pingTurn_ == false) {
+            ss << "pong" << repetitionsCounter_ - 1 << '\n';
+            cout << ss.rdbuf();
+        }
+
+        ss << "Timeout!\n";
+        cout << ss.rdbuf();
+        exit(0);
+    }
+
     void ping() {
-        // TODO: implement me :)
+        stringstream ss;
+        
+        ss << "ping0 ";
+        cout << ss.rdbuf();
+
+        for(; repetitionsCounter_ < repetitions_; ++repetitionsCounter_) {
+            unique_lock<mutex> lockCout(coutMutex_);
+            pingOrPong_.wait(lockCout, [&]{return pingTurn_;});
+            
+            ss << "ping" << repetitionsCounter_ << ' ';
+            cout << ss.rdbuf();
+            
+            pingTurn_ = false;
+            pingOrPong_.notify_all();
+        }
     }
 
     void pong() {
-        // TODO: implement me :)
+        stringstream ss;
+
+        for(int i = 0; i < repetitions_; ++i) {
+            unique_lock<mutex> lockCout(coutMutex_);
+            pingOrPong_.wait(lockCout, [&]{return !pingTurn_;});
+            
+            ss << "pong" << i << '\n';
+            cout << ss.rdbuf();
+            
+            pingTurn_ = true;
+            pingOrPong_.notify_all();
+        }
+        
+        safePongExit(ss);
     }
 
     void stop([[maybe_unused]] chrono::seconds timeout) {
-        // TODO: should stop execution after timeout
+        auto startCountTime = chrono::steady_clock::now();
+        
+        for(;;) {
```

Comment:
można użyć pętli `while(true)`

Reply:
Jasne, można. Tylko czy to jest preferowane użycie (while), czy po prostu dajesz znać, że jest coś takiego? Bo dla mnie to nie ma róźnicy. Akurat podpasowało mi tu for.

Reply:
Obie są ok, ale bardziej natrualnym zapisem jest `while(true)` niż for bez żadnych warunków. Ale w sumie zamiast tej pętli (która robi aktywne czekanie i grzeje procesor) oraz ifów niżej to lepiej użyć condition_variable pingOrPong_ z odpowiednimi warunkami.